### PR TITLE
This adds a Helm release validation tests for the Helm charts.

### LINKED ISSUE #241: There are helm release validation tests for the Conjur cluster prep helm charts
Helm release validation tests will make use of the [Helm Test feature](https://helm.sh/docs/helm/helm_test/) to validate deployed instances (releases) of Helm charts. 

Helm release validation tests are available for both the Kubernetes cluster prep Helm chart and the application Namespace prep Helm chart. 

The list of automated test cases below include both positive test cases (for which Helm test success is expected) and negative test cases (for which Helm test failure is expected, e.g. for incorrect configuration). In essence, we are “validating the validator” for our Helm release validation tests. 

The overall workflow for this automated testing is: 
- Create a Kubernetes cluster for testing. (This can be done using Kubernetes-in-Docker, a.k.a. KinD, in a GitHub action). 
- Create a Conjur instance. (This can be done using Conjur OSS Helm chart). 
- Generate and load Conjur policy for the authn-k8s authenticator. 
- Run `git clone …` to get a local copy of the Kubernetes cluster prep Helm chart. 
- Generate and load application-specific Conjur policy. 
- Run `git clone …` to get a local copy of the application Namespace prep Helm chart. 
- For each test case listed below: 
   - Create an appropriate, custom `values.yaml` file for the Kubernetes cluster prep Helm chart based on the test case scenario. 
   - Run `helm install …` for the Kubernetes cluster prep Helm chart using the modified `values.yaml`. 
   - Run `helm test` for the target Helm chart (second column in the chart below). 
   - Verify `helm test` results are as expected for this test case. 
   - Run `helm delete …` for Kubernetes cluster prep Helm chart. 

## Kubernetes cluster prep helm chart test scenarios
- [ ] Happy path: With correct chart values, Helm test passes (test Pod can authenticate with Conjur) 
- [ ] Incorrect Conjur URL: Helm test fails with indication that it cannot curl Conjur 
- [ ] Incorrect Conjur SSL certificate: Helm test fails with indication that configured Conjur SSL cert does not match Conjur’s actual SSL cert. 
- [ ] Incorrect Authenticator ID: Helm test fails with indication that the test Pod cannot authenticate with Conjur

### PR DESCRIPTION
**Pull request by rpothier**
### What does this PR do?
This adds a Helm release validation tests.
This script will create a Kind cluster, run the Helm cluster and namespace
prep Helm charts and run Helm tests.
There is also negative tests to validate the Helm tests fail as expected.

### What ticket does this PR close?
Resolves #241

#### Change log
- [ ] The CHANGELOG has been updated, or
- [x] This PR does not include user-facing changes and doesn't require a CHANGELOG update

#### Test coverage
- [ ] This PR includes new unit and integration tests to go with the code changes, or
- [x] The changes in this PR do not require tests

#### Documentation
- [ ] Docs (e.g. `README`s) were updated in this PR, and/or there is a follow-on issue to update docs, or
- [x] This PR does not require updating any documentation

#### Manual tests
**If you are preparing for a release**, have you run the following manual tests to verify existing functionality continues to function as expected?
- [ ] Manually run [Kubernetes-Conjur demo](https://github.com/conjurdemos/kubernetes-conjur-demo) with a local authn-k8s client image build

### CODE REVIEW

File: .github/workflows/ci.yaml [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+
+      - name: Install Helm
+        uses: azure/setup-helm@v1
+        with:
+          version: v3.3.0
+
+      - name: Install Helm
+        uses: azure/setup-helm@v1
+        with:
+          version: v3.2.1
```
Comment (diverdane) [inline .github/workflows/ci.yaml:37]: I think we can skip install of Helm 3.2.1, since 3.3.0 is also installed?

File: helm/conjur-config-cluster-prep/templates/tests/_helm_test.bats.txt [hunk 2]
```diff
@@ -50,7 +50,7 @@ function conjur_access_token_exists() {
   run conjur_is_reachable
   if [ "$status" -ne 0 ]; then
     display_error "The 'conjur.applianceURL' chart value is set to\n" \
```
Comment (diverdane) [inline helm/conjur-config-cluster-prep/templates/tests/_helm_test.bats.txt:52]: Thanks for cleaning up line 53. If you're making any other changes, the `applianceURL` should also be changed to `applianceUrl` on line 52.

File: bin/validate-helm [hunk 3]
```diff
@@ -0,0 +1,185 @@
...
+}
+
+function delete_cluster() {
+
+    kind delete cluster
+}
+
+readonly APPLIANCE_URL="https://conjur-oss.conjur-oss.svc.cluster.local"
+
+function git_clone() {
```
Comment (diverdane) [inline bin/validate-helm:49]: Looks like the `git_clone()` function clones the authn-k8s repo AND gets the Conjur cert. Should this function be renamed to reflect that its main purpose is to get the Conjur cert?

File: bin/validate-helm [hunk 4]
```diff
@@ -0,0 +1,185 @@
...
+    rm -rf conjur-authn-k8s-client
+    git clone https://github.com/cyberark/conjur-authn-k8s-client.git
+    pushd conjur-authn-k8s-client/helm/conjur-config-cluster-prep
+    ./bin/get-conjur-cert.sh -u $APPLIANCE_URL -i -v -s
+}
```
Comment (diverdane) [inline bin/validate-helm:54]: Should there be a `popd` here in `git_clone()`? i.e. have the function restore state as much as it can when it has done its thing?

File: bin/validate-helm [hunk 5]
```diff
@@ -0,0 +1,185 @@
...
+
+  pushd test-workflow
+
+  echo $(pwd)
+
+  if [ "$create_kind" = true ] ; then
+    create_cluster
+  fi
+  mkdir -p temp_verify
+  pushd temp_verify > /dev/null
```
Comment (diverdane) [inline bin/validate-helm:153]: The redirect to `/dev/null` can be removed, if `pushd` is calling your helper function that's doing the same.

File: bin/validate-helm [hunk 6]
```diff
@@ -0,0 +1,185 @@
...
+       banner "$RED" "Helm install Failed"
+  fi
+  helm get values "$helm_release"
+}
+
+function test_helm_chart() {
+
+  release=$1
+  test_name=$2
+  expected=1
```
Comment (diverdane) [inline bin/validate-helm:101]: Can we be certain that `helm test ...` will always return a 1 upon failure, or just that it will be non-zero?
Might take a few more conditionals, but would be safer to check for zero vs. non-zero?

**Review by diverdane — COMMENTED** [review]
Looking good!!! A few minor suggestions/comments.

File: helm/conjur-config-cluster-prep/templates/tests/_helm_test.bats.txt [hunk 7]
```diff
@@ -50,7 +50,7 @@ function conjur_access_token_exists() {
...
-                  "$conjurApplianceURL. This is not reachable via 'curl -k'"
+                  "$conjurApplianceUrl. This is not reachable via 'curl -k'"
```
Comment (doodlesbykumbi) [inline helm/conjur-config-cluster-prep/templates/tests/_helm_test.bats.txt:53]: I think we should consider updating the messaging to something that gives some explanation for the reachability  
```
This is not reachable outside the Kubernetes network, so running 'curl -k ...' on your local machine probably won't work!
```

File: helm/conjur-config-cluster-prep/templates/tests/_helm_test.bats.txt (hunk 2 above)
Comment (rpothier) [reply helm/conjur-config-cluster-prep/templates/tests/_helm_test.bats.txt:52]: Changed line 52 also.

File: helm/conjur-config-cluster-prep/templates/tests/_helm_test.bats.txt (hunk 7 above)
Comment (rpothier) [reply helm/conjur-config-cluster-prep/templates/tests/_helm_test.bats.txt:53]: It could be internal to the K8s network, and just a config error. That's what we are looking to detect.

File: .github/workflows/ci.yaml (hunk 1 above)
Comment (rpothier) [reply .github/workflows/ci.yaml:37]: yes, we don't need to downgrade.

File: bin/validate-helm (hunk 6 above)
Comment (rpothier) [reply bin/validate-helm:101]: good idea.

File: helm/conjur-config-cluster-prep/templates/tests/_helm_test.bats.txt (hunk 7 above)
Comment (diverdane) [reply helm/conjur-config-cluster-prep/templates/tests/_helm_test.bats.txt:53]: To add to that... the `curl -k <Conjur-appliance-URL>` is always executed from a Pod inside the K8s cluster because this is a Helm test, but the destination Conjur appliance URL can be an internal K8s address or an external-to-k8s address.

**Review by diverdane — APPROVED** [review]
